pkg/webhook/bd_replacement.go

```diff
@@ -0,0 +1,409 @@
+			if err != nil {
+				return nil, err
+			}
+			if oldNodeName == newNodeName {
+				commonPoolSpecs.oldSpec = append(commonPoolSpecs.oldSpec, oldPool)
+				commonPoolSpecs.newSpec = append(commonPoolSpecs.newSpec, newPool)
+				break
+			}
+		}
+	}
+	return commonPoolSpecs, nil
+}
+
+// IsReplacementValid validates the pool specs on CSPC for block device replacement case.
+func (bdr *BlockDeviceReplacement) IsReplacementValid(oldPoolSpec, newPoolSpec *apis.PoolSpec) (bool, string) {
+	newToOldBd := make(map[string]string)
+	for _, oldRg := range oldPoolSpec.RaidGroups {
+		oldRg := oldRg // pin it
+		if oldRg.Type == "" {
+			oldRg.Type = oldPoolSpec.PoolConfig.DefaultRaidGroupType
+		}
+		for _, newRg := range newPoolSpec.RaidGroups {
+			newRg := newRg // pin it
+			if IsRaidGroupCommon(oldRg, newRg) {
+				if IsBlockDeviceReplacementCase(&oldRg, &newRg) {
+					if ok, msg := bdr.IsBDReplacementValid(&newRg, &oldRg); !ok {
+						return false, msg
+					}
+					newBD := GetNewBDFromRaidGroups(&newRg, &oldRg)
+					for k, v := range newBD {
+						newToOldBd[k] = v
+					}
+				}
+				break
+			}
+		}
+	}
+
+	for newBD, oldBD := range newToOldBd {
+		err := createBDC(newBD, oldBD, bdr.OldCSPC)
+		if err != nil {
+			return false, err.Error()
+		}
+	}
+	return true, ""
+}
+
+// IsRaidGroupCommon returns true if the provided raid groups are the same raid groups.
+func IsRaidGroupCommon(rgOld, rgNew apis.RaidGroup) bool {
+	oldBdMap := make(map[string]bool)
+	for _, bd := range rgOld.BlockDevices {
+		oldBdMap[bd.BlockDeviceName] = true
+	}
+
+	for _, bd := range rgNew.BlockDevices {
+		if oldBdMap[bd.BlockDeviceName] {
+			return true
+		}
+	}
+	return false
+}
+
+// IsBlockDeviceReplacementCase returns true if the edit/update of CSPC can trigger a blockdevice
+// replacement.
+func IsBlockDeviceReplacementCase(newRaidGroup, oldRaidGroup *apis.RaidGroup) bool {
+	count := GetNumberOfDiskReplaced(newRaidGroup, oldRaidGroup)
+	return count >= 1
+}
+
+// GetNumberOfDiskReplaced returns the nuber of disk replaced in raid group.
+func GetNumberOfDiskReplaced(newRG, oldRG *apis.RaidGroup) int {
+	var count int
+	oldBlockDevicesMap := make(map[string]bool)
+	for _, bdOld := range oldRG.BlockDevices {
+		oldBlockDevicesMap[bdOld.BlockDeviceName] = true
+	}
+	for _, newBD := range newRG.BlockDevices {
+		if !oldBlockDevicesMap[newBD.BlockDeviceName] {
+			count++
+		}
+	}
+	return count
+}
+
+// IsBDReplacementValid validates for BD replacement.
+func (bdr *BlockDeviceReplacement) IsBDReplacementValid(newRG, oldRG *apis.RaidGroup) (bool, string) {
+
+	if oldRG.Type == string(apis.PoolStriped) {
+		return false, "cannot replace  blockdevice in stripe raid group"
+	}
+
+	// Not more than 1 bd should be replaced in a raid group.
+	if IsMoreThanOneDiskReplaced(newRG, oldRG) {
+		return false, "cannot replace more than one blockdevice in a raid group"
+	}
+
+	// The incoming BD for replacement should not pe present in the current CSPC.
```

Comment:
nit: `should be`

Reply:
The comment is good. So the newer BD that will be put in place should not be the BD that is already in CSPC ( in use by pool )

Reply:
It's about typo `pe-->be` ☺

Reply:
Oh. ok

Reply:
done